### PR TITLE
[ORTModule] Inline module-local functions before building gradient graph

### LINKED ISSUE #9375: ORT execution fails when a gradient builder is not registered for module-local functions
I'm trying to run a simple test case with ORTModule that has one linear layer inside a module as shown below:

```python
import torch
from onnxruntime.training.ortmodule import ORTModule

class M(torch.nn.Module):
      def __init__(self):
            super().__init__()
            self.fc1 = torch.nn.Linear(2, 2)
      def forward(self, x: torch.Tensor) -> torch.Tensor:
            return self.fc1(x)

x = torch.randn(2, 2)

# Case 1: This works
model = ORTModule(M())
y_pred = model(x)

# Case 2: This doesn't work
class ORTModuleExtension(ORTModule):
    def __init__(self, module, debug_options=None):
        super().__init__(module, debug_options)
        for training_mode in [False, True]:
            self._torch_module._execution_manager(training_mode)._export_extra_kwargs = {'export_modules_as_functions': True}

model = ORTModuleExtension(M())
y_pred = model(x)
```

In case 1 above, the ONNX graph exported by PyTorch has the module completely inlined:
case 1 graph

Whereas in case 2, it is wrapped inside a "flattened module" as shown below (the right image below is the graph corresponding to the function body of `_FlattenedModule`):
case 2 main graphcase 2 flattenedmodule

The latter case should ideally work since modules/functions are supposed to be inlined during partitioning. But I see the error below in case 2:
```
File "onnxruntime/training/ortmodule/_graph_execution_manager.py", line 232, in _build_graph
self._graph_builder.build()
RuntimeError: onnxruntime/orttraining/orttraining/core/graph/gradient_builder_registry.cc:28 onnxruntime::training::GradientDef onnxruntime::training::GetGradientForOp(const onnxruntime::training::GradientGraphConfiguration&, onnxruntime::Graph*, const onnxruntime::Node*, const std::unordered_set<std::__cxx11::basic_string<char> >&, const std::unordered_set<std::__cxx11::basic_string<char> >&, const onnxruntime::logging::Logger&, std::unordered_set<std::__cxx11::basic_string<char> >&)
gradient_builder != nullptr was false. The gradient builder has not been registered: _FlattenedModule for node _FlattenedModule_4
```

**System information**
ORT version: master
Torch version: Branch [onnx_ms_1](https://github.com/pytorch/pytorch/tree/onnx_ms_1)

### PR DESCRIPTION
**Pull request by adk9**
**Description**:

* Inline function nodes in the graph before building the gradient graph in ORTModule

**Motivation and Context**
This fixes #9375. Module-local functions are inlined during optimization of the graph before gradient building. If the graph was modified, i.e. if any functions were inlined, the transformers are run again on the inlined graph until there are no more functions to inline.

### CODE REVIEW

File: include/onnxruntime/core/graph/graph.h [hunk 1]
```diff
@@ -1017,6 +1017,13 @@ class Graph {
   */
   Node& FuseSubGraph(const IndexedSubGraph& sub_graph, const std::string& fused_node_name);
 
+  /**
+  Inline nodes in the graph that have an ONNX function definition but no matching ORT kernel
+  @param modified_graph Flag indicating whether the graph had any function nodes that were inlined
+  @returns Status indicating success or providing an error message.
+  */
+  Status InlineNodes(bool& modified_graph);
```
Comment (fninaparavecino) [inline include/onnxruntime/core/graph/graph.h:1025]: Why this can't be part of the graph_partitioner?

File: orttraining/orttraining/core/framework/ortmodule_graph_builder.cc [hunk 2]
```diff
@@ -78,14 +79,28 @@ Status OrtModuleGraphBuilder::Build(const std::vector<std::vector<int64_t>>* inp
...
+    if (!config_.build_gradient_graph) {
+      // This graph will be used only for inferencing, so stop right here.
+      // No need to build a gradient graph.
+      return Status::OK();
+    }
+
+    // expand any nodes that have an ONNX function definition but no matching ORT kernel.
+    modified_graph = false;
+    Graph& graph = gradient_model_->MainGraph();
+    ORT_RETURN_IF_ERROR(InlineNodes(graph, modified_graph));
```
Comment (souptc) [inline orttraining/orttraining/core/framework/ortmodule_graph_builder.cc:97]: is there any case that we define the gradient builder for a function? Instead of blindly inline everything, it is better we check the gradient builder registry for each node first, and only inline the nodes that doesn't have gradient builder registered.

Comment (adk9) [reply orttraining/orttraining/core/framework/ortmodule_graph_builder.cc]: > is there any case that we define the gradient builder for a function?
Yes, it could happen in case of custom ops where an EP claims a function and may have a gradient builder registered for it.
Instead of inlining all functions, I revised my PR to check if a) no EP claims the node and b) no gradient builder is registered for the node, before inlining the function node.

File: orttraining/orttraining/core/framework/ortmodule_graph_builder.cc [hunk 3]
```diff
@@ -63,6 +63,33 @@ Status OrtModuleGraphBuilder::Initialize(std::istream& model_istream,
   return Status::OK();
 }
 
+// Inline functions that do not have a gradient builder registered
+Status OrtModuleGraphBuilder::InlineFunctions(Graph& graph, bool& modified_graph) const {
+  // recurse into nested graphs first so we process from bottom up
```
Comment (souptc) [inline orttraining/orttraining/core/framework/ortmodule_graph_builder.cc:68]: i actually even not sure does ORT training support subgraph today. do you need that support?

**Comment by souptc** [comment]
@askhade , FYI, this is a PR to improve ORT gradient builder to support function nodes. In general it looks good to me. Could you help to review it?

**Comment by snnn** [comment]
This pull request has been automatically closed because it has merge conflicts and has been inactive for more than 30 days. Please rebase on the target branch and open a new PR.
